osquery/core/scope_guard.h

```diff
@@ -0,0 +1,58 @@
+/**
+ *  Copyright (c) 2018-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed as defined on the LICENSE file found in the
+ *  root directory of this source tree.
+ */
+
+#pragma once
+
+#include <utility>
+
+namespace osquery {
+
+/**
+ * The RAII based scope guard class.
+ *
+ * To be sure that resources are always released/removed/closed/verified/stoped
+ * in face of multiple return statements from the function.
+ *
+ * It takes functor object by value during the construction. It is going to be
+ * called once and only once during the destruction of ScopeGuard object.
+ *
+ * There is helper function to create the object of guard.
+ * @code{.cpp}
+ *   {
+ *     auto const manager = ScopeGuard<>::create(
+ *       [&file_path]() { fs::remove(file_path); }
+ *     );
+ *       ...
+ *     // it will be removed at the end of scope
+ *   }
+ * @endcode
+ */
+template <typename...>
+class ScopeGuard final {
```

Comment:
A free function is usually used for factory methods that perform type deduction for the caller. This allows you to not introduce a confusing template with a variadic template parameter pack for the class that seemengly only needs one template parameter.

Things like `std::make_pair()`, `std::make_tuple()` may be used as examples.

One way to do it would be to use a `ScopeGuard` namespace:

```cpp
namespace ScopeGuard
{

template <typename FinalRoutineType>
class ScopeGuardImpl
{
    // code from current ScopeGuard<FinalRoutineType> implementation
};

template <typename FinalRoutineType>
inline auto make(FinalRoutineType&& final_routine)
{
    return ScopeGuard<FinalRoutineType>(std::forward<FinalRoutineType>(final_routine));
}

} // namespace ScopeGuard
```

Since people will use `auto` when they use this type anyway, the API will look nice and clean from the outside while still being non-surprising from the inside (no variadic packs).

Reply:
The plan was to have all parts of scope guard under one "namespace". For it I could use either:
```c++
ScopeGuard<>::create(...);
```
or
```c++
scope_guard::create(...);
```
Ok, I got you point about overcomplicating with variadic templates.

Besides, the `std::make_pair()`, `std::make_tuple()` is not a perfect examples. Because the names consist two entities "action" and short "object" name. From some perspective connection between `std::unique_ptr` and `std::make_unique` is not completely obvious.

 Thank you, @SergNikitin !